docs/proposals/20191016-e2e-test-framework.md

```diff
@@ -0,0 +1,199 @@
+replaces: []
+superseded-by: []
+---
+
+# Cluster API testing framework
+
+## Table of Contents
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+
+- [Glossary](#glossary)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals/Future Work](#non-goalsfuture-work)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [As a Cluster API provider implementor I want to know my provider behaves in a way that Cluster API expects](#as-a-cluster-api-provider-implementor-i-want-to-know-my-provider-behaves-in-a-way-that-cluster-api-expects)
+    - [As a developer in the Cluster API ecosystem I want to have confidence that my change does not break expected behaviors](#as-a-developer-in-the-cluster-api-ecosystem-i-want-to-have-confidence-that-my-change-does-not-break-expected-behaviors)
+    - [As a Cluster API developer, I want to be sure I’m not accidentally disrupting existing providers.](#as-a-cluster-api-developer-i-want-to-be-sure-im-not-accidentally-disrupting-existing-providers)
+    - [As the release manager of a Cluster API provider, I want to be able to recommend new releases to users as soon as they come up](#as-the-release-manager-of-a-cluster-api-provider-i-want-to-be-able-to-recommend-new-releases-to-users-as-soon-as-they-come-up)
+    - [As a user of Cluster API, I want to be sure new releases I’m installing won’t cause regressions.](#as-a-user-of-cluster-api-i-want-to-be-sure-new-releases-im-installing-wont-cause-regressions)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Behaviors to test](#behaviors-to-test)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Alternatives](#alternatives)
+- [Upgrade Strategy](#upgrade-strategy)
+- [Additional Details](#additional-details)
+  - [Test Plan](#test-plan)
+  - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+- [Implementation History](#implementation-history)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+If this proposal adds new terms, or defines some, make the changes to the book's glossary when in PR stage.
+
+Basic e2es:
+The Reconcilers are exercised by creating a cluster and deleting a cluster
+Comprehensive e2es:
+Many of the components and behaviors of Cluster API are exercised.
+Creates a number of clusters and tears them down.
+
+## Summary
+
+Cluster API's providers could benefit from having a set of generic behavioral e2e tests. 
+
+| infra provider | has basic e2es | has comprehensive e2e suite |
+| --- |:---:|:---:|
+| CAPD | ❌ | ❌ |
+| CAPA | ✅ | ❌ |
+| CAPV | ❌ | ❌ |
+| CAPG | ✅ | ❌ |
+
+| bootstrap provider | has basic e2es | has comprehensive e2e suite |
+| --- |:---:|:---:|
+| CABPK| ❌ | ❌ |
```

Comment:
Should this be separate now that we're baking it into the main binary? Ideally I would like if we could helpers and maintain e2e style independence here, but I'm not sure if there has been thoughts on how to do that in the new model.

Reply:
removed.